src/main/java/io/cryostat/rules/RuleProcessor.java

```diff
@@ -120,91 +118,82 @@ public synchronized void disable() {
+            case LOST:
+                deactivate(null, tde.getServiceRef());
+                break;
+            default:
+                throw new UnsupportedOperationException(tde.getEventKind().toString());
         }
-        if (!EventKind.FOUND.equals(tde.getEventKind())) {
-            throw new UnsupportedOperationException(tde.getEventKind().toString());
+    }
+
+    private void activate(Rule rule, ServiceRef serviceRef) {
+        this.logger.trace(
+                "Activating rule {} for target {}", rule.getName(), serviceRef.getServiceUri());
+
+        Credentials credentials =
+                credentialsManager.getCredentials(serviceRef.getServiceUri().toString());
+        try {
+            startRuleRecording(new ConnectionDescriptor(serviceRef, credentials), rule);
+        } catch (Exception e) {
+            logger.error(e);
         }
-        registry.getRules(tde.getServiceRef())
-                .forEach(
-                        rule -> {
-                            this.logger.trace(
-                                    "Activating rule {} for target {}",
-                                    rule.getName(),
-                                    tde.getServiceRef().getServiceUri());
 
-                            Credentials credentials =
-                                    credentialsManager.getCredentials(
-                                            tde.getServiceRef().getServiceUri().toString());
-                            try {
-                                Future<Boolean> success =
-                                        startRuleRecording(
-                                                tde.getServiceRef().getServiceUri(),
-                                                rule.getRecordingName(),
-                                                rule.getEventSpecifier(),
-                                                rule.getMaxSizeBytes(),
-                                                rule.getMaxAgeSeconds(),
-                                                credentials);
-                                if (!success.get()) {
-                                    logger.trace("Rule activation failed");
-                                    return;
-                                }
-                            } catch (InterruptedException | ExecutionException e) {
-                                logger.error(e);
-                            }
+        logger.trace("Rule activation successful");
+        if (rule.getPreservedArchives() <= 0 || rule.getArchivalPeriodSeconds() <= 0) {
+            return;
+        }
+        tasks.put(
+                Pair.of(serviceRef, rule),
+                scheduler.scheduleAtFixedRate(
+                        periodicArchiverFactory.create(
+                                serviceRef, credentialsManager, rule, this::archivalFailureHandler),
+                        rule.getArchivalPeriodSeconds(),
+                        rule.getArchivalPeriodSeconds(),
```

Comment:
Does an archival period greater than 0 mean a rule is activated (i.e. recording started) once every period, after the previous recording (of the same rule) has been stopped and archived?

Reply:
Recordings are always started immediately using the template specified by the rule. The recording is never stopped, but periodic copies of it are made and preserved. `archivalPeriodSeconds` dictates how often these copies are made, `preservedArchives` dictates how many copies remain in storage (the oldest ones are pruned from storage when the limit is reached).